File: sdk/go/governance.go
// Package r3mes provides governance and staking operations for the R3MES network.
package r3mes

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
)

// Proposal represents a governance proposal.
type Proposal struct {
	ProposalID       string    `json:"proposal_id"`
	Content          any       `json:"content"`
	Status           string    `json:"status"`
	FinalTallyResult Tally     `json:"final_tally_result"`
	SubmitTime       string    `json:"submit_time"`
	DepositEndTime   string    `json:"deposit_end_time"`
	TotalDeposit     []Balance `json:"total_deposit"`
	VotingStartTime  string    `json:"voting_start_time"`
	VotingEndTime    string    `json:"voting_end_time"`
}

// Tally represents vote tally results.
type Tally struct {
	Yes        string `json:"yes"`
	Abstain    string `json:"abstain"`
	No         string `json:"no"`
	NoWithVeto string `json:"no_with_veto"`
}

// Vote represents a vote on a proposal.
type Vote struct {
	ProposalID string `json:"proposal_id"`
	Voter      string `json:"voter"`
	Option     string `json:"option"`
}

// GetProposals retrieves governance proposals.
func (c *Client) GetProposals(ctx context.Context, status string, limit int) ([]Proposal, error) {
	url := fmt.Sprintf("%s/cosmos/gov/v1beta1/proposals?pagination.limit=%d", c.config.RESTEndpoint, limit)
	if status != "" {
		url += fmt.Sprintf("&proposal_status=%s", status)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to execute request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("request failed with status %d: %s", resp.StatusCode, string(body))
	}

	var result struct {
		Proposals []Proposal `json:"proposals"`
	}

	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	return result.Proposals, nil
}

// GetProposal retrieves a specific proposal by ID.
func (c *Client) GetProposal(ctx context.Context, proposalID string) (*Proposal, error) {
	url := fmt.Sprintf("%s/cosmos/gov/v1beta1/proposals/%s", c.config.RESTEndpoint, proposalID)

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to execute request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusNotFound {
		return nil, ErrProposalNotFound
	}
	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("request failed with status %d: %s", resp.StatusCode, string(body))
	}

	var result struct {
		Proposal Proposal `json:"proposal"`
	}

	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	return &result.Proposal, nil
}

// GetProposalTally retrieves the current tally for a proposal.
func (c *Client) GetProposalTally(ctx context.Context, proposalID string) (*Tally, error) {
	url := fmt.Sprintf("%s/cosmos/gov/v1beta1/proposals/%s/tally", c.config.RESTEndpoint, proposalID)

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to execute request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("request failed with status %d: %s", resp.StatusCode, string(body))
	}

	var result struct {
		Tally Tally `json:"tally"`
	}

	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	return &result.Tally, nil
}

// GetValidatorsForGovernance retrieves validators (uses Validator type from blockchain.go).
func (c *Client) GetValidatorsForGovernance(ctx context.Context, status string, limit int) ([]Validator, error) {
	url := fmt.Sprintf("%s/cosmos/staking/v1beta1/validators?status=%s&pagination.limit=%d",
		c.config.RESTEndpoint, status, limit)

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to execute request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("request failed with status %d: %s", resp.StatusCode, string(body))
	}

	var result struct {
		Validators []struct {
			OperatorAddress string `json:"operator_address"`
			Description     struct {
				Moniker string `json:"moniker"`
			} `json:"description"`
			Tokens     string `json:"tokens"`
			Status     string `json:"status"`
			Jailed     bool   `json:"jailed"`
			Commission struct {
				CommissionRates struct {
					Rate string `json:"rate"`
				} `json:"commission_rates"`
			} `json:"commission"`
		} `json:"validators"`
	}

	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	validators := make([]Validator, len(result.Validators))
	for i, v := range result.Validators {
		validators[i] = Validator{
			OperatorAddress: v.OperatorAddress,
			Moniker:         v.Description.Moniker,
			Tokens:          v.Tokens,
			Status:          v.Status,
			Jailed:          v.Jailed,
			Commission:      v.Commission.CommissionRates.Rate,
		}
	}

	return validators, nil
}

// Delegation represents a delegation.
type Delegation struct {
	DelegatorAddress string  `json:"delegator_address"`
	ValidatorAddress string  `json:"validator_address"`
	Shares           string  `json:"shares"`
	Balance          Balance `json:"balance"`
}

// GetDelegations retrieves delegations for a delegator.
func (c *Client) GetDelegations(ctx context.Context, delegatorAddress string, limit int) ([]Delegation, error) {
	url := fmt.Sprintf("%s/cosmos/staking/v1beta1/delegations/%s?pagination.limit=%d",
		c.config.RESTEndpoint, delegatorAddress, limit)

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to execute request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("request failed with status %d: %s", resp.StatusCode, string(body))
	}

	var result struct {
		DelegationResponses []struct {
			Delegation struct {
				DelegatorAddress string `json:"delegator_address"`
				ValidatorAddress string `json:"validator_address"`
				Shares           string `json:"shares"`
			} `json:"delegation"`
			Balance Balance `json:"balance"`
		} `json:"delegation_responses"`
	}

	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	delegations := make([]Delegation, len(result.DelegationResponses))
	for i, d := range result.DelegationResponses {
		delegations[i] = Delegation{
			DelegatorAddress: d.Delegation.DelegatorAddress,
			ValidatorAddress: d.Delegation.ValidatorAddress,
			Shares:           d.Delegation.Shares,
			Balance:          d.Balance,
		}
	}

	return delegations, nil
}

// StakingPool represents the staking pool.
type StakingPool struct {
	BondedTokens    string `json:"bonded_tokens"`
	NotBondedTokens string `json:"not_bonded_tokens"`
}

// GetStakingPool retrieves the staking pool information.
func (c *Client) GetStakingPool(ctx context.Context) (*StakingPool, error) {
	url := fmt.Sprintf("%s/cosmos/staking/v1beta1/pool", c.config.RESTEndpoint)

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to execute request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("request failed with status %d: %s", resp.StatusCode, string(body))
	}

	var result struct {
		Pool StakingPool `json:"pool"`
	}

	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	return &result.Pool, nil
}

// Rewards represents delegation rewards.
type Rewards struct {
	Rewards []struct {
		ValidatorAddress string    `json:"validator_address"`
		Reward           []Balance `json:"reward"`
	} `json:"rewards"`
	Total []Balance `json:"total"`
}

// GetRewards retrieves delegation rewards for a delegator.
func (c *Client) GetRewards(ctx context.Context, delegatorAddress string) (*Rewards, error) {
	url := fmt.Sprintf("%s/cosmos/distribution/v1beta1/delegators/%s/rewards",
		c.config.RESTEndpoint, delegatorAddress)

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to execute request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("request failed with status %d: %s", resp.StatusCode, string(body))
	}

	var rewards Rewards
	if err := json.NewDecoder(resp.Body).Decode(&rewards); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	return &rewards, nil
}

File: sdk/python/r3mes/governance.py
"""
R3MES Governance Client

Client for governance operations including proposals and voting.
"""

import os
from typing import Optional, Dict, Any, List
import aiohttp

from .errors import (
    R3MESError,
    ConnectionError,
    NotFoundError,
    RateLimitError,
    TimeoutError,
)


class GovernanceClient:
    """
    Client for R3MES governance operations.

    Example:
        async with GovernanceClient() as gov:
            proposals = await gov.get_proposals()
            for p in proposals:
                print(f"{p['id']}: {p['title']}")
    """

    def __init__(
        self,
        rest_url: Optional[str] = None,
        backend_url: Optional[str] = None,
        timeout: float = 30.0,
    ):
        """
        Initialize the governance client.

        Args:
            rest_url: Cosmos REST API endpoint URL
            backend_url: R3MES backend service URL
            timeout: Request timeout in seconds
        """
        self.rest_url = rest_url or os.getenv(
            "R3MES_REST_URL", "https://api.r3mes.network"
        )
        self.backend_url = backend_url or os.getenv(
            "R3MES_BACKEND_URL", "https://backend.r3mes.network"
        )
        self.timeout = aiohttp.ClientTimeout(total=timeout)
        self._session: Optional[aiohttp.ClientSession] = None

    async def __aenter__(self) -> "GovernanceClient":
        self._session = aiohttp.ClientSession(timeout=self.timeout)
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb) -> None:
        await self.close()

    async def close(self) -> None:
        if self._session:
            await self._session.close()
            self._session = None

    def _get_session(self) -> aiohttp.ClientSession:
        if self._session is None:
            self._session = aiohttp.ClientSession(timeout=self.timeout)
        return self._session

    async def get_proposals(
        self,
        status: Optional[str] = None,
        limit: int = 100,
    ) -> List[Dict[str, Any]]:
        """
        Get governance proposals.

        Args:
            status: Filter by status ('voting', 'passed', 'rejected', 'deposit')
            limit: Maximum number of proposals to return

        Returns:
            List of proposal dictionaries
        """
        session = self._get_session()
        params = {"pagination.limit": limit}
        if status:
            params["proposal_status"] = status

        url = f"{self.rest_url}/cosmos/gov/v1beta1/proposals"
        
        async with session.get(url, params=params) as response:
            if response.status != 200:
                raise R3MESError(f"Failed to get proposals: {response.status}")
            data = await response.json()
            return data.get("proposals", [])

    async def get_proposal(self, proposal_id: int) -> Dict[str, Any]:
        """
        Get a specific proposal by ID.

        Args:
            proposal_id: The proposal ID

        Returns:
            Proposal dictionary
        """
        session = self._get_session()
        url = f"{self.rest_url}/cosmos/gov/v1beta1/proposals/{proposal_id}"
        
        async with session.get(url) as response:
            if response.status == 404:
                raise NotFoundError(f"Proposal not found: {proposal_id}")
            if response.status != 200:
                raise R3MESError(f"Failed to get proposal: {response.status}")
            data = await response.json()
            return data.get("proposal", {})

    async def get_proposal_votes(
        self,
        proposal_id: int,
        limit: int = 100,
    ) -> List[Dict[str, Any]]:
        """
        Get votes for a proposal.

        Args:
            proposal_id: The proposal ID
            limit: Maximum number of votes to return

        Returns:
            List of vote dictionaries
        """
        session = self._get_session()
        url = f"{self.rest_url}/cosmos/gov/v1beta1/proposals/{proposal_id}/votes"
        params = {"pagination.limit": limit}
        
        async with session.get(url, params=params) as response:
            if response.status != 200:
                raise R3MESError(f"Failed to get votes: {response.status}")
            data = await response.json()
            return data.get("votes", [])

    async def get_proposal_tally(self, proposal_id: int) -> Dict[str, Any]:
        """
        Get current tally for a proposal.

        Args:
            proposal_id: The proposal ID

        Returns:
            Tally dictionary with yes, no, abstain, no_with_veto counts
        """
        session = self._get_session()
        url = f"{self.rest_url}/cosmos/gov/v1beta1/proposals/{proposal_id}/tally"
        
        async with session.get(url) as response:
            if response.status != 200:
                raise R3MESError(f"Failed to get tally: {response.status}")
            data = await response.json()
            return data.get("tally", {})

    async def get_gov_params(self) -> Dict[str, Any]:
        """
        Get governance parameters.

        Returns:
            Dictionary with voting, deposit, and tally params
        """
        session = self._get_session()
        
        params = {}
        for param_type in ["voting", "deposit", "tallying"]:
            url = f"{self.rest_url}/cosmos/gov/v1beta1/params/{param_type}"
            async with session.get(url) as response:
                if response.status == 200:
                    data = await response.json()
                    params[param_type] = data.get(f"{param_type}_params", {})
        
        return params


class StakingClient:
    """
    Client for R3MES staking operations.

    Example:
        async with StakingClient() as staking:
            validators = await staking.get_validators()
            for v in validators:
                print(f"{v['moniker']}: {v['tokens']} tokens")
    """

    def __init__(
        self,
        rest_url: Optional[str] = None,
        timeout: float = 30.0,
    ):
        """
        Initialize the staking client.

        Args:
            rest_url: Cosmos REST API endpoint URL
            timeout: Request timeout in seconds
        """
        self.rest_url = rest_url or os.getenv(
            "R3MES_REST_URL", "https://api.r3mes.network"
        )
        self.timeout = aiohttp.ClientTimeout(total=timeout)
        self._session: Optional[aiohttp.ClientSession] = None

    async def __aenter__(self) -> "StakingClient":
        self._session = aiohttp.ClientSession(timeout=self.timeout)
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb) -> None:
        await self.close()

    async def close(self) -> None:
        if self._session:
            await self._session.close()
            self._session = None

    def _get_session(self) -> aiohttp.ClientSession:
        if self._session is None:
            self._session = aiohttp.ClientSession(timeout=self.timeout)
        return self._session

    async def get_validators(
        self,
        status: str = "BOND_STATUS_BONDED",
        limit: int = 100,
    ) -> List[Dict[str, Any]]:
        """
        Get validators.

        Args:
            status: Validator status filter
            limit: Maximum number of validators to return

        Returns:
            List of validator dictionaries
        """
        session = self._get_session()
        url = f"{self.rest_url}/cosmos/staking/v1beta1/validators"
        params = {"status": status, "pagination.limit": limit}
        
        async with session.get(url, params=params) as response:
            if response.status != 200:
                raise R3MESError(f"Failed to get validators: {response.status}")
            data = await response.json()
            return data.get("validators", [])

    async def get_validator(self, operator_address: str) -> Dict[str, Any]:
        """
        Get a specific validator.

        Args:
            operator_address: Validator operator address

        Returns:
            Validator dictionary
        """
        session = self._get_session()
        url = f"{self.rest_url}/cosmos/staking/v1beta1/validators/{operator_address}"
        
        async with session.get(url) as response:
            if response.status == 404:
                raise NotFoundError(f"Validator not found: {operator_address}")
            if response.status != 200:
                raise R3MESError(f"Failed to get validator: {response.status}")
            data = await response.json()
            return data.get("validator", {})

    async def get_delegations(
        self,
        delegator_address: str,
        limit: int = 100,
    ) -> List[Dict[str, Any]]:
        """
        Get delegations for a delegator.

        Args:
            delegator_address: Delegator wallet address
            limit: Maximum number of delegations to return

        Returns:
            List of delegation dictionaries
        """
        session = self._get_session()
        url = f"{self.rest_url}/cosmos/staking/v1beta1/delegations/{delegator_address}"
        params = {"pagination.limit": limit}
        
        async with session.get(url, params=params) as response:
            if response.status != 200:
                raise R3MESError(f"Failed to get delegations: {response.status}")
            data = await response.json()
            return data.get("delegation_responses", [])

    async def get_unbonding_delegations(
        self,
        delegator_address: str,
        limit: int = 100,
    ) -> List[Dict[str, Any]]:
        """
        Get unbonding delegations for a delegator.

        Args:
            delegator_address: Delegator wallet address
            limit: Maximum number of unbonding delegations to return

        Returns:
            List of unbonding delegation dictionaries
        """
        session = self._get_session()
        url = f"{self.rest_url}/cosmos/staking/v1beta1/delegators/{delegator_address}/unbonding_delegations"
        params = {"pagination.limit": limit}
        
        async with session.get(url, params=params) as response:
            if response.status != 200:
                raise R3MESError(f"Failed to get unbonding delegations: {response.status}")
            data = await response.json()
            return data.get("unbonding_responses", [])

    async def get_staking_pool(self) -> Dict[str, Any]:
        """
        Get staking pool information.

        Returns:
            Dictionary with bonded_tokens and not_bonded_tokens
        """
        session = self._get_session()
        url = f"{self.rest_url}/cosmos/staking/v1beta1/pool"
        
        async with session.get(url) as response:
            if response.status != 200:
                raise R3MESError(f"Failed to get staking pool: {response.status}")
            data = await response.json()
            return data.get("pool", {})

    async def get_staking_params(self) -> Dict[str, Any]:
        """
        Get staking parameters.

        Returns:
            Dictionary with unbonding_time, max_validators, etc.
        """
        session = self._get_session()
        url = f"{self.rest_url}/cosmos/staking/v1beta1/params"
        
        async with session.get(url) as response:
            if response.status != 200:
                raise R3MESError(f"Failed to get staking params: {response.status}")
            data = await response.json()
            return data.get("params", {})

    async def get_rewards(self, delegator_address: str) -> Dict[str, Any]:
        """
        Get delegation rewards for a delegator.

        Args:
            delegator_address: Delegator wallet address

        Returns:
            Dictionary with rewards and total
        """
        session = self._get_session()
        url = f"{self.rest_url}/cosmos/distribution/v1beta1/delegators/{delegator_address}/rewards"
        
        async with session.get(url) as response:
            if response.status != 200:
                raise R3MESError(f"Failed to get rewards: {response.status}")
            return await response.json()
